### PR TITLE
Upgrades: Android 12, Androidx-test 1.4, and RN 66

### LINKED ISSUE #2817: Upgrade to AndroidX-Test 1.4.0
## Description

Detox currently relies on Espresso and sibling deps corresponding to the [AndroidX-Test 1.2.0](https://github.com/android/android-test/releases/tag/androidx-test-1.2.0) release. Need to upgrade to the most recent stable one - ~[AndroidX-Test 1.3.0](https://github.com/android/android-test/releases/tag/androidx-test-1.3.0)~ [AndroidX-Test 1.4.0](https://github.com/android/android-test/releases/tag/androidx-test-1.4.0).

### LINKED ISSUE #2899: Android 12 and Java 11
## Description

Need to validate compliance with the upcoming release of Android 12

### LINKED ISSUE #3033: Upgrade to RN 0.66 (Android)
## Description

Add/verify support for the React Native 0.66 release:

- [x] Check (and migrate) Detox test app on developer env
- [ ] Upgrade demo apps (demo-react-native, then check demo-react-native-jest and detox-react-native-detox-instruments)
- [x] Upgrade associated CI job config to always use 0.66
- [x] Update support table in main README.md

### LINKED ISSUE #3034: Upgrade to RN 0.66 (iOS)
## Description

Add/verify support for the React Native 0.66 release:

- [x] Check (and migrate) Detox test app on developer env
- [ ] Upgrade demo apps (demo-react-native, then check demo-react-native-jest and detox-react-native-detox-instruments)
- [ ] Upgrade associated CI job config to always use 0.66
- [x] Update support table in main README.md

### PR DESCRIPTION
**Pull request by jonathanmos**
Upgrades the android version to 12, androidx-test version to 1.4.0 and RN to 0.66.1. 
Fixes #2899  
Fixes #2817 
Fixes #3033
Fixes #3034

### CODE REVIEW

File: scripts/ci.android.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash -ex
 
 # Approve unapproved SDK licenses
-yes | $ANDROID_HOME/tools/bin/sdkmanager --licenses
```
Comment (mikehardy) [inline scripts/ci.android.sh:4]: oh that is subtle! it has bitten me before as well
mostly commenting to say that after giving this a really focused attempt in #3055 I can really sympathize with the difficulty here. The RN59 build failure was confounding for me, this is really touchy to support a react-native version skew as great as RN59 to RN67

Comment (jonathanmos) [reply scripts/ci.android.sh:4]: Yes, I had to make this change for it to work on the build agent with updated sdk. I think it's 99% of the way to a working pr, but there's still a couple of small issues remaining. Thanks for your reviews and feedback on this!

File: detox/test/scripts/postinstall.js [hunk 2]
```diff
@@ -47,6 +47,21 @@ function overrideReactAndroidGradleForRn64Android() {
   fs.copySync(PATCH_SCRIPT_PATH, REACT_ANDROID_GRADLE_SCRIPT_PATH);
 }
 
+function overrideReactAndroidGradleForRn66Android() {
```
Comment (d4vidi) [inline detox/test/scripts/postinstall.js:50]: @jonathanmos where is `'ReactAndroid_rn66_build.gradle`, though?

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java [hunk 3]
```diff
@@ -122,6 +140,33 @@ protected boolean checkIdle() {
         return true;
     }
 
+    // This is a workaround for https://github.com/facebook/react-native/issues/32594
+    // uses duck typing heuristics to determine that this is probably the stuck Switch operation and if so, ignores it
+    private boolean workaroundForRN66Bug(Object uiOperationQueue) {
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java:145]: This must be extracted onto a separate class/module. We cannot keep being affected by the broken-window effect of these legacy-code grade classes. Here's one example as to how this can be achieved: https://github.com/wix/Detox/blob/d95d924ae323edfab832f281b7713f6519139d0d/detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/SerialExecutorReflected.kt#L6

File: detox/android/build.gradle [hunk 4]
```diff
@@ -3,9 +3,9 @@ buildscript {
         isOfficialDetoxLib = true
         kotlinVersion = '1.2.0'
         dokkaVersion = '0.9.18'
-        buildToolsVersion = '29.0.3'
-        compileSdkVersion = 29
-        targetSdkVersion = 29
+        buildToolsVersion = '30.0.2'
```
Comment (d4vidi) [inline detox/android/build.gradle:6]: I know we talked about this but seems like I need you to refresh my memory... why not use build tools 31.x.x?

File: detox/android/detox/build.gradle [hunk 5]
```diff
@@ -80,18 +80,18 @@ dependencies {
...
 dependencies {
 
-    // Versions are in-sync with the 'androidx-test-1.2.0' release/tag of the android-test github repo,
-    // used by the Detox generator. See https://github.com/android/android-test/releases/tag/androidx-test-1.2.0
+    // Versions are in-sync with the 'androidx-test-1.4.0' release/tag of the android-test github repo,
+    // used by the Detox generator. See https://github.com/android/android-test/releases/tag/androidx-test-1.4.0
     // Important: Should remain so when generator tag is replaced!
-    api('androidx.test.espresso:espresso-core:3.3.0') { // Needed all across Detox but also makes Espresso seamlessly provided to Detox users with hybrid apps/E2E-tests.
+    api('androidx.test.espresso:espresso-core:3.4.0') { // Needed all across Detox but also makes Espresso seamlessly provided to Detox users with hybrid apps/E2E-tests.
         exclude group: 'com.google.code.findbugs', module: 'jsr305'
```
Comment (d4vidi) [inline detox/android/detox/build.gradle:90]: This may no longer be needed, in fact

File: detox/android/detox/build.gradle [hunk 6]
```diff
@@ -80,18 +80,18 @@ dependencies {
...
     }
-    api 'androidx.test.espresso:espresso-web:3.3.0' // Web-View testing
-    api 'androidx.test:rules:1.2.0' // Needed because of ActivityTestRule. Needed by users *and* internally used by Detox.
-    api 'androidx.test.ext:junit:1.1.1' // Needed so as to seamlessly provide AndroidJUnit4 to Detox users. Depends on junit core.
+    api 'androidx.test.espresso:espresso-web:3.4.0' // Web-View testing
+    api 'androidx.test:rules:1.4.0' // Needed because of ActivityTestRule. Needed by users *and* internally used by Detox.
+    api 'androidx.test.ext:junit:1.1.3' // Needed so as to seamlessly provide AndroidJUnit4 to Detox users. Depends on junit core.
```
Comment (d4vidi) [inline detox/android/detox/build.gradle:94]: Consider boy-scouting by replacing the comments with a `because` (example in line 104 below - https://github.com/wix/Detox/pull/3035/files#diff-ef16074b7489954843e57dfef5376da6115f35ea0b6be19781edcb542530437bR104)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java (hunk 3 above)
Comment (d4vidi) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java:145]: Double ping :-)

File: detox/test/index.android.js [hunk 7]
```diff
@@ -10,4 +10,5 @@ class exampleAndroid extends example {
 
 }
 
+console.disableYellowBox = true;
```
Comment (d4vidi) [inline detox/test/index.android.js:13]: Let's not push this 😅

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java [hunk 8]
```diff
@@ -122,6 +140,33 @@ protected boolean checkIdle() {
...
+        boolean isStuckSwitchOperation = false;
+
+        try {
+            Class<?> DispatchCommandViewOperation = Class.forName(CLASS_DISPATCH_COMMAND_VIEW_OPERATION);
+            List<?> mViewCommandOperations = Reflect.on(uiOperationQueue).field(FIELD_MVIEW_COMMAND_OPERATIONS).get();
+
+            if (mViewCommandOperations.size() == 1) {
+                Class<?> ReadableArray = Class.forName(CLASS_READABLE_ARRAY);
+                Object viewOperation = DispatchCommandViewOperation.cast(mViewCommandOperations.get(0));
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java:154]: Why is this `.cast()` needed?

Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java:154]: Could you verify that it is a `DispatchStringCommandOperation` class, specifically?

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java [hunk 9]
```diff
@@ -122,6 +140,33 @@ protected boolean checkIdle() {
...
+
+                Object viewArgs = ReadableArray.cast(Reflect.on(viewOperation).field(FIELD_MARGS).get());
+                Method getReadableArrayBoolean = viewArgs.getClass().getDeclaredMethod(METHOD_GET_BOOLEAN, int.class);
+                getReadableArrayBoolean.setAccessible(true);
+                Boolean nativeValueArg = (Boolean) getReadableArrayBoolean.invoke(viewArgs, 0);
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java:159]: I think `Reflect.on()` does this for you and therefor is usable here as well

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java [hunk 10]
```diff
@@ -122,6 +140,33 @@ protected boolean checkIdle() {
...
+
+                int numRetries = Reflect.on(viewOperation).field(FIELD_NUM_RETRIES).get();
+                String mCommand = Reflect.on(viewOperation).field(FIELD_MCOMMAND).get();
+
+                isStuckSwitchOperation = (numRetries == 1 && mCommand.equals(SET_NATIVE_VALUE) && nativeValueArg != null);
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIModuleIdlingResource.java:164]: So this is pretty good! 💪🏻💪🏻💪🏻  But since false-idles is an insanely sensitive issue, I must ask for two improvements over it:

1. I noticed that using the operation's `mTag` field, there is a way to reach the actual Android view instance (!), provided we can get a hold of the `NativeViewHierarchyManager` instance: property `NativeViewHierarchyManager.mTagsToViews` is a `tag-->view` map. Ultimately, we need to assert that the view's class type is indeed the naughty switch that's causing all the issues.
1. Let's not conclude "idle" immediately in the first occurrence of things. Rather, since as you said -- the operation will be stuck forever, we can double check that this isn't just a bad coincidence (and thus avoid a false-idle...) by stabilizing the check such that we'd only declare idle if all of the above conditions are met X times in row (X can be even as high a 10, IMO).

With that, I'd get some better sleep at night 😄

**Review by d4vidi — CHANGES_REQUESTED** [review]
Brilliant work but unfortunately I must make our lives even harder in this case... Please go over my notes, kind sir!

File: detox/android/build.gradle (hunk 4 above)
Comment (mikehardy) [reply detox/android/build.gradle:6]: build tools will fail unless a user is on gradle 7 toolchain (gradle and android gradle plugin) which I don't think is a guarantee https://stackoverflow.com/questions/68387270/android-studio-error-installed-build-tools-revision-31-0-0-is-corrupted (there's a google issuetracker bug as well, they appear to have no intention of fixing it after months). 30 is thus more compatible for now, surely

I have personally spent my time on this, total waste ;-) https://github.com/invertase/notifee/issues/211

File: detox/test/scripts/postinstall.js (hunk 2 above)
Comment (jonathanmos) [reply detox/test/scripts/postinstall.js:50]: [it's in the same path](https://github.com/wix/Detox/tree/android-12/detox/test/scripts)

File: detox/android/build.gradle (hunk 4 above)
Comment (d4vidi) [reply detox/android/build.gradle:6]: Understood. Thanks!

File: detox/android/detox/build.gradle [hunk 11]
```diff
@@ -80,30 +80,40 @@ dependencies {
...
 }
 
 // Third-party/extension deps.
 dependencies {
     implementation('com.google.android.material:material:1.2.1') {
         because 'Material components are mentioned explicitly (e.g. Slider in get-attributes handler)'
     }
-    implementation 'org.apache.commons:commons-lang3:3.7' // Needed by invoke. Warning: Upgrading to newer versions is not seamless.
+    implementation('org.apache.commons:commons-lang3:3.7') {
+        because 'Needed by invoke. Warning: Upgrading to newer versions is not seamless.'
```
Comment (d4vidi) [inline detox/android/detox/build.gradle:115]: 😻

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        boolean isStuckSwitchOperation = false;
+
+        try {
+            Class<?> DispatchStringCommandOperation = Class.forName(CLASS_DISPATCH_COMMAND_OPERATION);
+            List<?> mViewCommandOperations = Reflect.on(uiOperationQueue).field(FIELD_MVIEW_COMMAND_OPERATIONS).get();
+
+            if (mViewCommandOperations.size() == 1) {
+                Object viewOperation = DispatchStringCommandOperation.cast(mViewCommandOperations.get(0));
+
+                int mTag = Reflect.on(viewOperation).field(FIELD_MTAG).get();
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java:38]: naming convention (no `m` prefixes)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    private int timesStuckQueueDetected = 0;
+
+    // This is a workaround for https://github.com/facebook/react-native/issues/32594
+    // uses duck typing heuristics to determine that this is probably the stuck Switch operation and if so, ignores it
+    boolean workaroundForRN66Bug(Object uiOperationQueue) {
+        boolean isStuckSwitchOperation = false;
+
+        try {
+            Class<?> DispatchStringCommandOperation = Class.forName(CLASS_DISPATCH_COMMAND_OPERATION);
+            List<?> mViewCommandOperations = Reflect.on(uiOperationQueue).field(FIELD_MVIEW_COMMAND_OPERATIONS).get();
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java:33]: naming convention (no `m` prefixes)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                if (timesStuckQueueDetected >= NUM_TIMES_BEFORE_NOTIFY_IDLE) {
+                    isStuckSwitchOperation = true;
+                }
+            } else {
+                timesStuckQueueDetected = 0;
+            }
+        } catch (ClassNotFoundException e) {
+            Log.d(LOG_TAG, String.valueOf(e.getCause()));
+        }
+        return isStuckSwitchOperation;
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java:63]: This is a step in the right direction, but it still requires more work.

1. The point behind `XXXReflected` classes is to provide an easier access into the inner fields using reflection (i.e. abstract away the ugly reflection details). This concept is still not utilized well enough in this case. We need something more in the likes of `UIManagerModuleReflected`, `NativeHierarchyManagerReflected` and `ViewOperationReflected` classes, for that. Once applied, we must have the reflection details and the workaround logic *separated* into different classes / methods.
1. I think that detecting whether _an_ "anonymous" suspicious command is repeatedly stuck in the queue is okay, but it can and should be made more robust: We should try harder and see whether _the_ suspicious command is stuck, by repeatedly comparing whether the current command is identical to our temp (weak) reference to it.

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+                int mTag = Reflect.on(viewOperation).field(FIELD_MTAG).get();
+                UIViewOperationQueue uiManagerModule = (UIViewOperationQueue) uiOperationQueue;
+                NativeViewHierarchyManager nativeHierarchyManager = Reflect.on(uiManagerModule).field(FIELD_MVIEW_MANAGERS).get();
+                String mClassName = nativeHierarchyManager.resolveView(mTag).getClass().getSimpleName();
+
+                int numRetries = Reflect.on(viewOperation).field(FIELD_NUM_RETRIES).get();
+                String mCommand = Reflect.on(viewOperation).field(FIELD_MCOMMAND).get();
+
+                boolean isReactSwitch = mClassName.equals(REACT_SWITCH);
+                boolean hasOneRetryIncremented = numRetries == 1;
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java:47]: Why is this needed?

File: detox/test/scripts/postinstall.js (hunk 2 above)
Comment (d4vidi) [reply detox/test/scripts/postinstall.js:50]: Could you please add the `// Edited by Detox' post-install script` note to it?

**Review by d4vidi — CHANGES_REQUESTED** [review]
Almost there... @jonathanmos please review

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+                int mTag = Reflect.on(viewOperation).field(FIELD_MTAG).get();
+                UIViewOperationQueue uiManagerModule = (UIViewOperationQueue) uiOperationQueue;
+                NativeViewHierarchyManager nativeHierarchyManager = Reflect.on(uiManagerModule).field(FIELD_MVIEW_MANAGERS).get();
+                String mClassName = nativeHierarchyManager.resolveView(mTag).getClass().getSimpleName();
+
+                int numRetries = Reflect.on(viewOperation).field(FIELD_NUM_RETRIES).get();
+                String mCommand = Reflect.on(viewOperation).field(FIELD_MCOMMAND).get();
+
+                boolean isReactSwitch = mClassName.equals(REACT_SWITCH);
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java:46]: Also, let's not assume that the class _is_ `ReactSwitch`, but check that it is either exactly it or a child of it (i.e. inherits it)

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java (hunk 15 above)
Comment (jonathanmos) [reply detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java:47]: This indicates that the operation has been queued for retry, which happens when the reactSwitch operation fails in the rn66 bug

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIManagerModuleReflected.kt [hunk 17]
```diff
@@ -1,98 +0,0 @@
...
-        }
-    }
-
-    fun getNonBatchOpsEmpty(): Boolean {
-        val uiViewOperationQueueInstance = getInstance()
-        return if (uiViewOperationQueueInstance == null) true else {
-            synchronized(Reflect.on(uiViewOperationQueueInstance).field(LOCK_OPERATIONS).get())
-            {
-                Reflect.on(uiViewOperationQueueInstance)
-                    .field(FIELD_DISPATCH_RUNNABLES)
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/UIManagerModuleReflected.kt]: (should have been: `FIELD_NON_BATCHED_OPS_LOCK`)

**Comment by d4vidi** [comment]
@jonathanmos I took the liberty to push a commit (https://github.com/wix/Detox/pull/3035/commits/dbe0db4f5e8a3bf0ac0fa680e65c72f5ca5148f3) here in order to try to proactively assist in pushing this forward towards a merge.

I addressed several domains:
1. Mostly styling - tried to make things more Kotlin idiomatic
2. A touch of architecture: made Reflected classes more tightly isolated in associated with the class they are reflecting
3. Found a bug 🐞 ! (see previous comment)

**Review by d4vidi — APPROVED** [review]

**Comment by mikehardy** [comment]
@jonathanmos & @d4vidi Exciting to see this merged! Fantastic to see Detox all ready for another mobile version year (with iOS15 and Android12 thoroughly handled). Amazing to me that it requires so much effort some times to handle the versions. Hopefully Android13 and iOS16 won't be so difficult :-). Cheers

**Comment by d4vidi** [comment]
@mikehardy Thanks for the kind words, and for being involved in the process 🙏🏻
Unfortunately, as RN have been holding back the official version 1 release since forever, from time to time a version upgrade happens to be very painful. I truly hope this would not be the case in at least a couple of versions 🤞🏻

In any case, solving the worked-around issue is being handled on a separate thread, as I'm sure you already know.

**Comment by mikehardy** [comment]
Yes, I've been a part of the RN release process as well, working hard on getting Xcode 13+ and Apple Silicon / M1 chips handled there, it is *really* fiddly to get the compile working and took quite a few iterations. Then in the end a Hermes upgrade resulted in rejected builds from Apple. Sometimes it just takes a while but the current RN67 release candidate has no show-stoppers reported after around a week of testing amongst our group of pre-release testers (finally!) and I've had successful TestFlight builds with it on M1 macs even, so maybe finally it's good to go :-). Cheers

**Comment by d4vidi** [comment]
Best of luck! We've been reported of the hermes issues - happy to see it's coming together nonetheless. Cheers!

File: detox/test/ios/Podfile [hunk 18]
```diff
@@ -29,4 +29,5 @@ end
 
 post_install do |installer|
   react_native_post_install(installer)
+  __apply_Xcode_12_5_M1_post_install_workaround(installer)
```
Comment (asafkorem) [inline detox/test/ios/Podfile:32]: We are not using Xcode 12.5 + M1 on our agents 😠

Comment (mikehardy) [reply detox/test/ios/Podfile:32]: This is part of the standard template, I worked so hard on that workaround 😅 - I'm not sure why 12.5+ (or 13+ now?) isn't in use in the agents? But if it is at all you will suffer immediate build failures without this workaround, and in fact there may be build failures just by including Swift, not just using 12.5+ or M1. Either way, it's part of the standard template and will also be removed in react-native 0.68 (0.67 still requires it unfortunately...). Building on Xcode right now is a real mess basically, and I would remove that line at your peril...

Comment (asafkorem) [reply detox/test/ios/Podfile:32]: Thanks for all your efforts on this upgrade Mike. I just saw the workaround details on RN. We are using Xcode 13, so it is relevant for our Xcode version. I was too quick to judge that it was relevant for v12.5 and not for the following versions (or specifically v13)..

I saw this error in [one of our builds](https://buildkite.com/wix-mobile-oss/detox/builds/126#98f5e3da-1543-43a6-a9d9-18e4ebe49178), and wondered why we need this call on our Podfile if we are not using M1 agents:
```
Generating Pods project
--
  | [!] An error occurred while processing the post-install hook of the Podfile.
  |  
  | undefined method `__apply_Xcode_12_5_M1_post_install_workaround' for #<Pod::Podfile:0x00007f7f5bd58f68>
  |  
  | /Users/builder/work/detox/test/ios/Podfile:32:in `block (2 levels) in from_ruby'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-core-1.10.2/lib/cocoapods-core/podfile.rb:179:in `post_install!'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:897:in `run_podfile_post_install_hook'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:885:in `block in run_podfile_post_install_hooks'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:145:in `message'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:884:in `run_podfile_post_install_hooks'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:329:in `block (2 levels) in create_and_save_projects'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer/xcode/pods_project_generator/pods_project_writer.rb:61:in `write!'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:328:in `block in create_and_save_projects'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:64:in `section'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:307:in `create_and_save_projects'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:299:in `generate_pods_project'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:178:in `integrate'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:166:in `install!'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/command/install.rb:52:in `run'
  | /Library/Ruby/Gems/2.6.0/gems/claide-1.0.3/lib/claide/command.rb:334:in `run'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/lib/cocoapods/command.rb:52:in `run'
  | /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.2/bin/pod:55:in `<top (required)>'
  | /usr/local/bin/pod:23:in `load'
  | /usr/local/bin/pod:23:in `<main>'
```

However, when [tried to remove this line](https://github.com/wix/Detox/pull/3146), our build failed on `xcodebuild` command..

Comment (asafkorem) [reply detox/test/ios/Podfile:32]: OK, the issue was that we are calling this method from RN64 build.

Comment (mikehardy) [reply detox/test/ios/Podfile:32]: Oh! Yes, that won't work :-), great catch
